File: Go/Algorithmic_Blueprints/Sorting_and_Order_Statistics/counting_sort/counting_sort.go
package main

import (
	"fmt"
)

// CountingSort sorts an array of integers using the counting sort algorithm.
// Note: This implementation assumes integers are within a reasonable range (0 to max).
func CountingSort(arr []int) []int {
	if len(arr) == 0 {
		return arr
	}

	max := arr[0]
	for _, v := range arr {
		if v > max {
			max = v
		}
	}

	count := make([]int, max+1)
	for _, v := range arr {
		count[v]++
	}

	for i := 1; i <= max; i++ {
		count[i] += count[i-1]
	}

	output := make([]int, len(arr))
	for i := len(arr) - 1; i >= 0; i-- {
		output[count[arr[i]]-1] = arr[i]
		count[arr[i]]--
	}

	return output
}

func main() {
	sample := []int{4, 2, 2, 8, 3, 3, 1}
	fmt.Println("Original array:", sample)
	sorted := CountingSort(sample)
	fmt.Println("Sorted array:  ", sorted)
}

File: algorithms/sorting/counting_sort/counting_sort.py

def counting_sort(arr):
    """
    Sorts a list of non-negative integers using the Counting Sort algorithm.

    Args:
        arr (list): A list of non-negative integers.

    Returns:
        list: A new sorted list.
    """
    if not arr:
        return arr

    max_val = max(arr)
    count = [0] * (max_val + 1)

    # Count the frequency of each element
    for num in arr:
        count[num] += 1

    # Reconstruct the sorted array
    sorted_arr = []
    for value, freq in enumerate(count):
        sorted_arr.extend([value] * freq)

    return sorted_arr


def main():
    try:
        data = [2, 1, 4, 5, 3]

        sorted_data = counting_sort(data)
        print("Counting Sort:", sorted_data)

    except Exception as e:
        print("Exception:", e)


if __name__ == "__main__":
    main()

